ARCs/arc-1337.md

```diff
@@ -0,0 +1,170 @@
+    shares: RoyaltyShare[], // All the shares defined for this policy
+    allowed_assets: number[] // The list of asset ids that are acceptable for payment
+}
+
+```
+
+### Royalty Enforcer
+
+The Royalty Enforcer is an application that controls the transfer of ASAs subject to the Royalty Policy. 
+
+This is accomplished by exposing an interface defined as a set of [ABI Methods](#abi-methods) allowing a grouped transaction call 
+
+### Royalty Asset 
+
+The Royalty Asset is an ASA subject to royalty collection and MUST be created with the the appropriate roles and parameters. 
+
+> Because the protocol does not allow updating a role after it's been deleted, if the creator thinks she may want to modify them later, they must be set to some non-zero address.
+
+
+#### Asset Roles
+
+The Clawback role MUST be set to the Application Address.
+
+> Since the Royalty Enforcer relies on using the clawback mechanism to perform the transfer the Clawback should NEVER be set to the zero address. 
+
+The Freeze role MAY be set to zero OR the Application Address. 
+
+> If the creator wants to allow an ASA to be Royalty Free after some conditions are met, it should be set to the Application Address 
+
+The Manager role MAY be set to zero OR the Application Address.
+
+> If the creator wants to update the Freeze role, this should be set to the application address
+
+The Reserve role may be set to anything.
+
+#### Asset Parameters
+
+The DefaultFrozen address MAY be set to true or false.
+
+> In the case that a creator wants the asset to no longer be subject to royalty collection, this should be set to false.
+
+
+
+### ABI Methods
+
+
+*Required*
+
+
+SetPolicy
+    
+    set_policy(royalty_asset: asset, royalty_shares: Tuple(address, uint64)[], allowed_assets: asset[])
+
+Sets the Royalty Policy for a given asset. 
+
+The `asset` is the ASA id of the asset the policy is set for.
+
+The `royalty_shares` are encoded as an array of `tuple(address, uint64)` that represents the Royalty Recipient and Royalty Share in basis points. The address of a Royalty Recipient SHOULD NOT be repeated, only the last occurence of the address will be used for their Royalty Share. 
+
+> The maximum number of Royalty Recipients is limited to the number of foreign accounts that may be passed from a grouped set of application call transactions.  As of go-algorand v3.4.0 this number is:
+> 4 [max accts per call] * (16 [max calls per group] - 1 [transaction reserved for payment]) = 60
+
+The `allowed_assets` list contains the ASA ids that are acceptable for payment. The contract logic MUST opt into the set of assets specified in order to accept the payments as part of a transfer.
+
+> All the Royalty Recipients SHOULD be opted into the full set of assets contained in this list but MAY NOT be checked ahead of time. In the case that an account is not opted into an asset, any transfers where payment is specified for that asset will fail until the account opts into the asset.
```

Comment:
Quick sanity check: suppose I own a `Royalty ASA`, the royalty is payed with another `Share ASA`. If the `Royalty Collector` opts-out the `Share ASA` any transfer of `Royalty ASA` would fail. 

Should we have any countermeasure if this happens? 

Suggestion: If this happens I would expect that the `Royalty ASA` is now “royalty free” (since the `Royalty Collector` doesn't want to receive `Share ASA` at all). So if `Royalty Enforcer` looks at `Royalty Collector` and see that it opted-out the `Share ASA` then `move` method should be available to anyone.

WDYT?

Reply:
Hm, I think we should fail at the protocol level if this happens. It may be the case that the `Royalty Collector` no longer wants to accept that asset and they've just not updated the `Royalty Policy` yet. 

If we default to move then accidentally opting out of the asset will make them all royalty free until the `Royalty Collector` opts back in or the `Royalty Policy` is updated.  Unless someone is watching closely they might not notice this condition if we allow the transaction to succeed by moving but they'd definitely notice if folks were complaining about failed transactions.

There is a broader point here where if some pre-condition fails folks aren't able to trade the asset anymore, which I think _should_ have more thought put into it.

Reply:
> There is a broader point here where if some pre-condition fails folks aren't able to trade the asset anymore, which I think _should_ have more thought put into it.

Exactly, the corner case I want to avoid (and be able to handle somehow) is the "lock-in effect" by which people are no longer able to transfer and trade their ASA not because they refuse to comply with the Royalty Policy but because that Policy is broken by the creator ex-post.